udp/udp.go

```diff
@@ -14,47 +18,98 @@ const (
 	udpBufferSize = 65536
 )
 
+var (
+	// ErrBindAddressRequired is returned when starting the Server
+	// without a TCP or UDP listening address.
+	ErrBindAddressRequired = errors.New("bind address required")
+)
+
 // SeriesWriter defines the interface for the destination of the data.
 type SeriesWriter interface {
-	WriteSeries(database, retentionPolicy string, points []influxdb.Point) (uint64, error)
+	WriteSeries(string, string, []influxdb.Point) (uint64, error)
 }
 
-// UDPServer
 type UDPServer struct {
 	writer SeriesWriter
+	conn   *net.UDPConn
+	addr   *net.UDPAddr
+	wg     sync.WaitGroup
+
+	batchSize    int
+	batchTimeout time.Duration
+
+	Logger *log.Logger
+
+	host string
 }
 
 // NewUDPServer returns a new instance of a UDPServer
 func NewUDPServer(w SeriesWriter) *UDPServer {
 	u := UDPServer{
 		writer: w,
+		Logger: log.New(os.Stderr, "[udp] ", log.LstdFlags),
 	}
 	return &u
 }
 
-// ListenAndServe binds the server to the given UDP interface.
-func (u *UDPServer) ListenAndServe(iface string) error {
-	addr, err := net.ResolveUDPAddr("udp", iface)
+func (u *UDPServer) SetBatchSize(sz int)             { u.batchSize = sz }
+func (u *UDPServer) SetBatchTimeout(d time.Duration) { u.batchTimeout = d }
+
+// ListenAndServer instructs the UDPServer to start processing data
+// on the given interface. iface must be in the form host:port.
+func (u *UDPServer) ListenAndServe(iface string) (err error) {
+	if iface == "" { // Make sure we have an address
+		return ErrBindAddressRequired
+	}
+
+	u.addr, err = net.ResolveUDPAddr("udp", iface)
 	if err != nil {
 		log.Printf("Failed resolve UDP address %s: %s", iface, err)
 		return err
 	}
 
-	conn, err := net.ListenUDP("udp", addr)
+	u.conn, err = net.ListenUDP("udp", u.addr)
 	if err != nil {
-		log.Printf("Failed set up UDP listener at address %s: %s", addr, err)
+		log.Printf("Failed set up UDP listener at address %s: %s", u.addr, err)
 		return err
 	}
 
 	var bp client.BatchPoints
 	buf := make([]byte, udpBufferSize)
 
+	u.host = u.addr.String()
+
+	batcher := influxdb.NewPointBatcher(u.batchSize, u.batchTimeout)
+	in, out := batcher.Start()
+
+	// Start processing batches.
+	var wg sync.WaitGroup
+	done := make(chan struct{})
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case batch := <-out:
+				_, e := u.writer.WriteSeries(bp.Database, bp.RetentionPolicy, batch)
+				if e != nil {
+					u.Logger.Printf("failed to write point batch to database %q: %s\n", bp.Database, e)
+				}
+			case <-done:
+				return
+			}
+		}
+	}()
+
+	u.wg.Add(1)
 	go func() {
+		defer u.wg.Done()
 		for {
-			_, _, err := conn.ReadFromUDP(buf)
+			_, _, err := u.conn.ReadFromUDP(buf)
 			if err != nil {
-				log.Printf("Failed read UDP message: %s.", err)
-				continue
+				close(done)
```

Comment:
We just added a new `Flush` call to the batcher, which you should add here, just before the call to `close`.